Framework: Remove selected site reset in boot

tyxla opened the pull request:
This PR removes the code that resets the selected site before executing each route. Previously, this was necessary to fix a glitch with site selection (see 7633-gh-calypso-pre-oss and 8102-gh-calypso-pre-oss).

However, the resetting no longer appears to be necessary. If we remove it, it will take care of edge cases when we go from a site-specific page to a non-site specific page (like #8215). 

This PR also removes the now obsolete `SitesList.prototype.resetSelectedSite`.

Fixes #8215.

To test:

### Test case 1
* Navigate to My Sites
* Confirm that a site is selected from the sidebar site picker
* Navigate to Reader
* Click New Post in the master bar
* Verify that the site selector popover is shown

### Test case 2
* In a fresh Calypso page session, navigate to My Sites
* Select a site if one isn't already selected
* Click your profile image to the right of the master bar (i.e. `/me`)
* Verify that the current My Sites view maintains the selected site until after `/me` loads

### Test case 3
* Ensure you are logged in to wp.com
* In wp-admin for a disconnected, non-premium Jetpack site, click "Connect Jetpack"
* Click 'Approve'
* On the plans page, click any element in the top nav, eg 'Reader' or profile link
* Verify you're directed to the Reader and everything works smoothly without any crashes.

/cc @aduth @johnHackworth

aduth commented:
I've not tested yet, but my initial reaction is: We should have some mechanism in application state such that there's not a currently selected site when viewing pages like Reader. Whether that's middleware, or Redux selected state responding to `ROUTE_SET`, or responding to `SECTION_SET` when navigating from a site-specific section to a non-site-specific section. I suspect the problem in #8215 is probably due to some component assuming that a selected site exists. We should instead be guarded about making these assumptions.

aduth commented:
>there's not a currently selected site when viewing pages like Reader.

Playing devil's advocate with myself, maybe this _is_ desirable such that we preserve the site which had been selected while managing sites. This is the current behavior anyways, though I don't recall off the top of my head how this is implemented.

aduth commented on client/boot/index.js:
```diff
@@ -359,17 +359,6 @@ function reduxStoreReady( reduxStore ) {
-		// occur after the sections routes execute to avoid a brief flash where
-		// sites are reset but the next section is waiting to be loaded.
-		if ( ! route.getSiteFragment( context.path ) && sites.getSelectedSite() ) {
-			sites.resetSelectedSite();
```
Removing this line makes `SitesList.prototype.resetSelectedSite` unused. It should probably be removed too.

tyxla replied:
Didn't notice - thanks!

aduth commented:
If I recall correctly, the primary motivation for resetting the site was related to previous logic in the editor which would launch the editor directly if there was a selected site. This was later changed to always show the site picker when clicking the New Post button. Unless there's another element of `<Layout />` that changes its behavior based on the presence of a selected site, I think we could probably be safe in removing this.

tyxla commented:
Feel free to add any test cases to the above ones. I think this is ready for review anyway, cc @aduth @johnHackworth @roccotripaldi

aduth commented:
I guess we're still using selected site to assign the `href` to the New Post button here:

https://github.com/Automattic/wp-calypso/blob/ab552d23f64c7ce447dc640e1312c4983a8d4bec/client/layout/masterbar/publish.jsx#L73

The default behavior is prevented though, so this only has an effect when, for example, right-clicking and opening as new tab. This might be a bit unexpected when trying to start a new post in a new tab while viewing the reader (should it open to site selector? or to the site that had previous been selected?)

cc also @mtias for thoughts on keeping site selected through navigating to different sections of the app.

mtias commented:
I believe some area of Me/Help was also using last site selected.

tyxla commented:
@mtias you're right, it's used in `help/contact` to determine the last selected site for the contact form. 

So, let me speculate a little. What would change if we set the `lastSelected` site like we do in `resetSelectedSite()`, but we don't touch `selected` and leave it as is? This would still fix #8215 but would not add a regression to `/help/contact`. @aduth thoughts?

aduth commented:
@tyxla I'm not totally clear on what it is you're suggesting.

If we were to remove site resetting in this pull request, couldn't we just change `help/contact` to call `sites.getSelectedSite()` since the last selected site would still be selected at that point?

(If we do change `help/contact` away from using `getLastSelectedSite`, we should also remove that function from `SitesList`, since that's the only place it's used)

tyxla commented:
@aduth that's even better - I've got rid of `getLastSelectedSite` completely and replaced it with `getSelectedSite()`, which appears to work great in all of the discussed cases when we're not resetting the site. Care to take another look?

aduth commented:
This looks good and works well for me 👍 

I did some digging to figure out how the My Sites link is preserved after navigating away. Appears there's a separate [`lib/site-stats-sticky-tab` module](https://github.com/Automattic/wp-calypso/tree/master/client/lib/site-stats-sticky-tab) for that. We should seek to migrate that to Redux at some point, but should have no effect on the changes here.

mtias commented:
Trying to figure out why the all-sites view is completely unresponsive to me in this branch.

aduth commented:
>Trying to figure out why the all-sites view is completely unresponsive to me in this branch.

Is there a specific screen or steps leading up to All Sites that leads it to become unresponsive? I'm not able to reproduce any such issue myself.

mtias commented:
From any single site while stats, switching to all-sites kicks the CPU to over 100% and I can't navigate.

tyxla commented:
> From any single site while stats, switching to all-sites kicks the CPU to over 100% and I can't navigate.

Odd, I'm also unable to reproduce. And it happens only in this branch?

mtias commented:
Yes.

mtias commented:
Regression: if I'm on an all sites views (`/posts`) then navigate to "Me" and click on My Sites again I'm returned to previously selected site, not the all sites view.

aduth commented:
Sounds like we might need to reset the selected site when choosing All Sites view?

mtias commented:
It seems this is a more significant change that affects the concept of "selected site" in non-site views — I think we need to put some more thought into it.

tyxla commented:
@mtias I'm unable to reproduce that regression. Could you please send more detailed steps (I might be missing a step)?

mtias commented:
- Load `http://calypso.localhost:3000/posts`.
- Click "Me" in masterbar.
- Click "My Sites" in masterbar.

tyxla commented:
@mtias that was a good catch. It was a kind of an edge case where the selected site is not being taken into consideration. It was caused by the `siteStatsStickyTab` store, which does some work to persist the last selected site slug and filter (day, week, month, etc). 

However, for All sites it only needs to preserve the filter, and not the selected site slug.

I've now updated the PR to do some more:

* When selecting an All Sites view, reset the `siteStatsStickyTab` store as well, preserving the previous stats filter.
* When selecting an All Sites view, If the previous stats filter was `"insights"`, we default to `"day"`, because we don't have insights in All Sites stats view.

@mtias @aduth I could use some more eyes and feedback here, thank you.

aduth commented:
I find the `siteStatsStickyTabStore` logic to be very difficult to follow. Why do we pass empty string as slug instead of another falsey value (`null`, `undefined`, `false`)? I wonder how much effort it would be to migrate away from this store and use Redux-based state for generating the master bar link. Only thing we might not have currently is the stats filter (How important is this to preserve anyways? That said, many folks are very resistant to stats updates, so I don't want to be held accountable for suggesting a change 😆 )

All that said, the regression @mtias noted seems to be resolved in my testing.

tyxla commented:
@aduth I agree the `siteStatsStickyTabStore` isn't very pretty to use, but it's outside of the scope of this PR. Is there anything that holds us from merging this one? Can you confirm we no longer have the regression @mtias?

tyxla commented:
I've "kindly borrowed" `getCurrentSiteSlug()` from #8879 (thanks a ton @seear!) so we don't introduce another usage of `sites-list`'s `getSelectedSite()`. I've also implemented it into the Help contact form component, and fixed some ESLint warnings within it.

This is ready for a new review @aduth. And cc @seear as promised.

aduth commented:
Full steps to reproduce, since it seems intermittent:

1. Start at Reader
2. Click My Sites
3. Note selected site
4. Click Help in sidebar footer
5. Click Contact Us
6. Click My Sites
7. Switch site
8. Repeat steps 3-5
9. Observe that site selector is set to site noted in step 3, not in step 8

I've confirmed this is an issue in master too, so not related to the changes here. Nonetheless we should plan to address it in a subsequent pull request. In the meantime, I think we should be good to move forward with this then.

aduth commented:
Tracking bug at https://github.com/Automattic/wp-calypso/issues/8964
